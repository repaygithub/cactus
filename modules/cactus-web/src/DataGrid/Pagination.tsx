import { assignWith, isEqual } from 'lodash'
import PropTypes from 'prop-types'
import React from 'react'

import Pagination, { PaginationProps } from '../Pagination/Pagination'
import PrevNext, { PrevNextProps } from '../PrevNext/PrevNext'
import { useDataGridContext } from './DataGridContext'
import { PageState, PageStateAction, pageStateKeys, Pagisort } from './types'

interface PageStateProps extends PageState {
  initialPage?: number
}

export interface ExtendedPaginationProps
  extends Omit<PaginationProps, 'currentPage' | 'pageCount'>,
    PageStateProps {}

export interface ExtendedPrevNextProps extends PrevNextProps, PageStateProps {}

export const calcPageState = (pageState: PageState): PageState => {
  // TS appeasement: comparison operators work perfectly fine with undefined.
  const p = pageState as Required<PageState>
  if (p.pageSize > 0) {
    if (p.itemCount >= 0) {
      p.pageCount = Math.ceil(p.itemCount / p.pageSize)
    }
    if (p.currentPage > 0) {
      if (p.currentPage > p.pageCount) {
        p.currentPage = Math.max(1, p.pageCount)
      }
      p.itemOffset = (p.currentPage - 1) * p.pageSize
    } else if (p.itemOffset >= 0) {
      p.currentPage = Math.floor(p.itemOffset / p.pageSize) + 1
      if (p.currentPage > p.pageCount) {
        p.currentPage = Math.max(1, p.pageCount)
        p.itemOffset = (p.currentPage - 1) * p.pageSize
      }
    }
  }
  return pageState
}

const assignIfUndefined = (dest: unknown, src: unknown) => dest ?? src

const interleave = (state: PageState, updates: PageState) => {
  // To get the correct derived/implied values from the updates, we do the
  // calculations twice: first with updates + pageSize, then combined with state.
  const combinedState = calcPageState({ pageSize: state.pageSize, ...updates })
  return calcPageState(assignWith(combinedState, state, assignIfUndefined))
}

type PageStateReducer = React.Reducer<PageState, PageStateAction>
export const pageStateReducer: PageStateReducer = (state, action) => {
  const updates = typeof action === 'function' ? action(state) : action
  const nextState = interleave(state, updates)
  return isEqual(state, nextState) ? state : nextState
}

const usePageState = (props: PageStateProps & Pagisort) => {
  const { pageState, updatePageState } = useDataGridContext('DataGrid.PageState')
  const pagePropValues = []
  const controlledProps: PageState = {}
  for (const key of pageStateKeys) {
    pagePropValues.push(props[key])
    if (props[key] !== undefined) controlledProps[key] = props[key]
    delete props[key] // Remove from props so the rest can be safely forwarded.
  }
  // In case they just spread the entire "pagisort" object.
  delete props.sort
  React.useEffect(() => {
    updatePageState(controlledProps, false)
  }, pagePropValues) // eslint-disable-line react-hooks/exhaustive-deps

  const combinedState = interleave(pageState, controlledProps)
  if (!combinedState.currentPage) {
    controlledProps.currentPage = combinedState.currentPage = props.initialPage || 1
  }
  delete props.initialPage
  return [combinedState, updatePageState] as const
}

export const DataGridPagination: React.FC<ExtendedPaginationProps> = ({
  onPageChange,
  ...props
}) => {
  const [pageState, updatePageState] = usePageState(props)
  if (!pageState.currentPage || !pageState.pageCount) return null
  return (
    <Pagination
      {...props}
      currentPage={pageState.currentPage}
      pageCount={pageState.pageCount}
      onPageChange={(currentPage) => {
        updatePageState({ currentPage }, true)
        onPageChange?.(currentPage)
      }}
    />
  )
}
DataGridPagination.displayName = 'DataGrid.Pagination'

DataGridPagination.propTypes = pageStateKeys.reduce((pt, key) => {
  pt[key] = PropTypes.number
  return pt
}, {} as Record<keyof PageState, typeof PropTypes.number>)

const incrementPage: PageStateAction = ({ currentPage = 0 }) => ({ currentPage: currentPage + 1 })
const decrementPage: PageStateAction = ({ currentPage = 2 }) => ({ currentPage: currentPage - 1 })

export const DataGridPrevNext: React.FC<ExtendedPrevNextProps> = ({ onNavigate, ...props }) => {
  const [pageState, updatePageState] = usePageState(props)
  return (
    <PrevNext
      disablePrev={pageState.currentPage === 1}
      onNavigate={(direction: 'prev' | 'next'): void => {
        const changePage = direction === 'prev' ? decrementPage : incrementPage
        updatePageState(changePage, true)
        onNavigate?.(direction)
      }}
      {...props}
    />
  )
}
DataGridPrevNext.displayName = 'DataGrid.PrevNext'

DataGridPrevNext.propTypes = {
  currentPage: PropTypes.number,
  pageSize: PropTypes.number,
  itemOffset: PropTypes.number,
}
