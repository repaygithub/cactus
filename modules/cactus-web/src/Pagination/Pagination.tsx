import { border } from '../helpers/theme'
import { margin, MarginProps } from 'styled-system'
import {
  NavigationChevronLeft,
  NavigationChevronRight,
  NavigationMenuDots,
} from '@repay/cactus-icons'
import pick from 'lodash/pick'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

export interface PageLinkProps {
  'aria-label': string
  'aria-current'?: 'page'
  page: number
  disabled: boolean
  children: React.ReactChild
  rel?: string
  className?: string
  onClick?: () => void
}

interface PaginationProps extends MarginProps {
  pageCount: number
  currentPage: number
  className?: string
  onPageChange?: (page: number) => void
  linkAs?: React.ComponentType<PageLinkProps>
  label?: string
  currentPageLabel?: string
  prevPageLabel?: string
  nextPageLabel?: string
  lastPageLabel?: string
  makeLinkLabel?: (page: number) => string
}

interface CommonPageProps {
  currentPage: number
  className?: string
  'aria-label': string
  'aria-current'?: 'page'
  linkAs?: React.ComponentType<PageLinkProps>
  pageChangeHandler: (page: number) => (() => void) | undefined
}

interface PageProps extends CommonPageProps {
  page: number
  rel?: string
  children: React.ReactChild
}

const PageButton = (props: PageProps) => {
  const { page, currentPage, linkAs, pageChangeHandler, children, ...rest } = props
  const disabled = page === currentPage
  return (
    <PageItem>
      <PageLink
        page={page}
        as={linkAs}
        onClick={disabled ? undefined : pageChangeHandler(page)}
        disabled={disabled}
        {...rest}
      >
        {children}
      </PageLink>
    </PageItem>
  )
}

function getPageButton(page: number, props: CommonPageProps, label: string) {
  if (page === props.currentPage) {
    props = { ...props, 'aria-current': 'page' }
  } else {
    props = { ...props, 'aria-label': label }
  }
  return (
    <PageButton key={page} page={page} {...props}>
      {page}
    </PageButton>
  )
}

function dots(key: string) {
  return (
    <PageItem key={key}>
      <RotatedMenuDots />
    </PageItem>
  )
}

function getPages(
  commonProps: CommonPageProps,
  lastPage: number,
  makeLinkLabel: (page: number) => string
) {
  // TODO At some point we might want to make these breakpoints configurable,
  // or controlled by media queries.
  const lastBreak = lastPage - 2
  let lowerBreak = Math.max(2, commonProps.currentPage - 2)
  let upperBreak = lowerBreak + 5
  if (upperBreak >= lastBreak) {
    lowerBreak = lowerBreak + lastBreak - upperBreak
    upperBreak = lastPage - 1
  }
  if (lowerBreak <= 2) {
    lowerBreak = 1
  }

  const pages: Array<React.ReactElement> = []
  if (lowerBreak > 1) {
    pages.push(dots('lower-break'))
  }
  for (let page = lowerBreak; page <= upperBreak; page++) {
    pages.push(getPageButton(page, commonProps, makeLinkLabel(page)))
  }
  if (upperBreak <= lastBreak) {
    pages.push(dots('upper-break'))
  }
  pages.push(getPageButton(lastPage, commonProps, makeLinkLabel(lastPage)))
  return pages
}

const PageLinkBase: React.FC<PageLinkProps> = (props: PageLinkProps) => {
  const { page, disabled, children, onClick, ...rest } = props
  const linkProps: React.AnchorHTMLAttributes<HTMLAnchorElement> = rest
  if (disabled) {
    linkProps['aria-disabled'] = 'true'
  } else {
    linkProps.tabIndex = 0
  }
  if (onClick) {
    // TODO This is a handy accessibility tool, should we make it available?
    linkProps.onKeyPress = (event: React.KeyboardEvent) => {
      if (event.key === 'Enter' || event.key === ' ') {
        event.preventDefault()
        onClick()
      }
    }
  }
  return (
    <a role="link" onClick={onClick} {...linkProps}>
      {children}
    </a>
  )
}

export const Pagination: React.FC<PaginationProps> = (props) => {
  const { pageCount, currentPage, label, className, onPageChange, linkAs } = props
  const pageChangeHandler = React.useCallback(
    (page: number) => {
      if (onPageChange) {
        return () => onPageChange(page)
      }
    },
    [onPageChange]
  )

  if (pageCount < 1 || currentPage < 1 || currentPage > pageCount) {
    return null
  }

  // *sigh* I wish Typescript could play nicer with defaultProps...
  const makeLinkLabel = props.makeLinkLabel as (page: number) => string
  const lastPageLabel = `${props.lastPageLabel}, ${pageCount}`
  const commonProps = {
    linkAs,
    currentPage,
    pageChangeHandler,
    'aria-label': `${props.currentPageLabel}, ${currentPage}`,
  }
  const prevProps = { ...commonProps, page: Math.max(1, currentPage - 1) }
  prevProps['aria-label'] = `${props.prevPageLabel}, ${currentPage - 1}`
  const nextProps = { ...commonProps, page: Math.min(pageCount, currentPage + 1) }
  nextProps['aria-label'] = `${props.nextPageLabel}, ${currentPage + 1}`
  const marginProps = pick(props, margin.propNames as Array<string>)

  return (
    <Nav className={className} aria-label={label} {...marginProps}>
      <PageList role="list">
        <PageButton {...commonProps} page={1} aria-label={makeLinkLabel(1)}>
          <NavigationFirst />
        </PageButton>
        <PageButton {...prevProps} rel="prev">
          <NavigationChevronLeft />
        </PageButton>

        {getPages(commonProps, pageCount, makeLinkLabel)}

        <PageButton {...nextProps} rel="next">
          <NavigationChevronRight />
        </PageButton>
        <PageButton {...commonProps} page={pageCount} aria-label={lastPageLabel}>
          <NavigationLast />
        </PageButton>
      </PageList>
    </Nav>
  )
}

Pagination.displayName = 'Pagination'

//@ts-ignore
Pagination.propTypes = {
  pageCount: function (props: PaginationProps, propName, componentName) {
    const pageCount = parseInt(props.pageCount as any)
    if (pageCount < 1 || pageCount !== parseFloat(props.pageCount as any)) {
      return new Error('Prop `pageCount` must be a positive integer')
    }
  },
  currentPage: function (props: PaginationProps, propName, componentName) {
    const pageCount = parseInt(props.pageCount as any)
    const current = parseInt(props.currentPage as any)
    if (current < 1 || current > pageCount || current !== parseFloat(props.currentPage as any)) {
      return new Error('Prop `currentPage` must be an integer in the range [1, `pageCount`]')
    }
  },
  onPageChange: function (props: PaginationProps, propName, componentName) {
    if (!props.onPageChange) {
      if (!props.linkAs) {
        return new Error('Either `linkAs` OR `onPageChange` prop is required')
      }
    } else if (typeof props.onPageChange !== 'function') {
      return new Error('Prop `onPageChange` must be a function')
    }
  },
  linkAs: PropTypes.elementType,
  label: PropTypes.string.isRequired,
  currentPageLabel: PropTypes.string.isRequired,
  prevPageLabel: PropTypes.string.isRequired,
  nextPageLabel: PropTypes.string.isRequired,
  lastPageLabel: PropTypes.string.isRequired,
  makeLinkLabel: PropTypes.func.isRequired,
}

Pagination.defaultProps = {
  label: 'Pages',
  currentPageLabel: 'Current page',
  prevPageLabel: 'Go to previous page',
  nextPageLabel: 'Go to next page',
  lastPageLabel: 'Go to last page',
  makeLinkLabel: (page: number) => `Go to page ${page}`,
}

export default Pagination

// TODO Use the real icons once they're added to the library
const NavigationFirst = () => {
  return (
    <svg
      fill="currentcolor"
      width="16"
      height="16"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M2.66671 1.33335C2.66671 0.965164 2.36823 0.666687 2.00004 0.666687C1.63185 0.666687 1.33337 0.965164 1.33337 1.33335V14.6667C1.33337 15.0349 1.63185 15.3334 2.00004 15.3334C2.36823 15.3334 2.66671 15.0349 2.66671 14.6667L2.66671 1.33335ZM13.5286 0.861949C13.789 0.6016 14.2111 0.6016 14.4714 0.861949C14.7318 1.1223 14.7318 1.54441 14.4714 1.80476L8.27618 8.00002L14.4714 14.1953C14.7318 14.4556 14.7318 14.8777 14.4714 15.1381C14.2111 15.3984 13.789 15.3984 13.5286 15.1381L6.86197 8.47142C6.60162 8.21107 6.60162 7.78897 6.86197 7.52862L13.5286 0.861949Z" />
    </svg>
  )
}

const NavigationLast = () => {
  return (
    <svg
      fill="currentcolor"
      width="16"
      height="16"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M14.6667 1.33335C14.6667 0.965164 14.3682 0.666687 14 0.666687C13.6319 0.666687 13.3334 0.965164 13.3334 1.33335L13.3334 14.6667C13.3334 15.0349 13.6318 15.3334 14 15.3334C14.3682 15.3334 14.6667 15.0349 14.6667 14.6667L14.6667 1.33335ZM1.52864 0.861949C1.26829 1.1223 1.26829 1.54441 1.52864 1.80476L7.7239 8.00002L1.52864 14.1953C1.26829 14.4556 1.26829 14.8777 1.52864 15.1381C1.78899 15.3984 2.2111 15.3984 2.47144 15.1381L9.13811 8.47142C9.39846 8.21107 9.39846 7.78897 9.13811 7.52862L2.47144 0.861949C2.2111 0.6016 1.78899 0.6016 1.52864 0.861949Z" />
    </svg>
  )
}

const RotatedMenuDots = styled(NavigationMenuDots)`
  margin-top: 5px;
  transform: rotate(90deg);
  cursor: default;
  display: block;
`

const Nav = styled.nav`
  ${margin}
`

const PageList = styled.ol`
  display: flex;
  flex-wrap: nowrap;
  list-style: none;
  margin: 0;
  padding: 0;
`

const PageItem = styled.li`
  min-width: 15px;
  height: 24px;
  text-align: center;
  background: none;
  appearance: none;
  padding: 2px 8px;
  display: block;

  &,
  :link {
    color: ${(p) => p.theme.colors.darkestContrast};
    ${(p) => p.theme.textStyles.small};
    line-height: 18px;
    text-decoration: none;
  }

  border-left: ${(p) => border(p.theme, 'lightContrast')};

  &:last-child {
    border-right: ${(p) => border(p.theme, 'lightContrast')};
  }

  svg {
    width: 15px;
    height: 15px;
    vertical-align: middle;
  }
`

const PageLink = styled(PageLinkBase)<PageLinkProps>`
  cursor: pointer;
  padding: 3px;
  vertical-align: middle;
  line-height: 18px;
  display: block;
  border-radius: 8px;

  &:visited {
    color: ${(p) => p.theme.colors.mediumContrast};
  }

  // For styling active links, and overriding the :visited color.
  text-decoration: none;
  color: inherit;

  &:hover {
    color: ${(p) => p.theme.colors.callToAction};
  }

  &:active,
  &:focus {
    // Re-stated to prevent a small shift when the button is clicked in Firefox
    padding: 3px;
    outline: none;
    ${(p) => p.theme.colorStyles.callToAction};
  }

  // Just in case they want to use a button for 'linkAs'
  border: none;
  background-color: transparent;
  font: inherit;
  &::-moz-focus-inner {
    border: none;
  }

  &[disabled],
  &[aria-disabled='true'] {
    color: ${(p) => p.theme.colors.lightGray};
    background-color: transparent;
    cursor: default;
  }

  &[aria-current='page'] {
    ${(p) => p.theme.colorStyles.base};
  }
`
