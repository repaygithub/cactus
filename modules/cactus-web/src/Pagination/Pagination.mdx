---
name: Pagination
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import Pagination from './Pagination'
import cactusTheme from '@repay/cactus-theme'

# Pagination

The `Pagination` component creates an interactive list of page numbers which should be implemented to change the screen based on the current page.

## Best practices

The `Pagination` component requires size, current, and onPageChange props. The size needs to be set to how many pages there will be on the webpage. The current prop is used to set the current page value, and its initial value will be the initial page that is selected. The onPageChange should be passed the action that will occur when the user selects a page.

## Basic usage

```jsx
import React from 'react'
import Pagination from '@repay/cactus-web'

function ManagedPagination({ size }: any) {
  const [current, setCurrent] = React.useState(1)
  return <Pagination current={current} size={size} onPageChange={setCurrent} />
}

export default () => <ManagedPagination size={20} />
```

## Properties

<PropsTable of={Pagination} />
