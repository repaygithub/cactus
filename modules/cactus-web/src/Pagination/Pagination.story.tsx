import React from 'react'

import { Pagination } from '../'
import { Action, actions, HIDE_CONTROL, Story } from '../helpers/storybook'
import { PageLinkProps } from './Pagination'

export default {
  title: 'Pagination',
  component: Pagination,
  argTypes: {
    currentPage: HIDE_CONTROL,
    makeLinkLabel: HIDE_CONTROL,
    disabled: { control: 'boolean' },
    ...actions('onPageChange'),
  },
} as const

const stop = (handler: any) => (e: React.MouseEvent) => {
  e.preventDefault()
  handler()
}
const LinkWithHref = ({ page, disabled, onClick, ...props }: PageLinkProps) => (
  <a aria-disabled={disabled} href={`#${page}`} onClick={stop(onClick)} {...props} />
)

type PStory = Story<typeof Pagination, { onPageChange: Action<number> }>
export const BasicUsage: PStory = ({ onPageChange, ...args }) => {
  const [current, setCurrent] = React.useState(1)
  return <Pagination {...args} currentPage={current} onPageChange={onPageChange.wrap(setCurrent)} />
}
BasicUsage.args = { pageCount: 10, width: 'inherit' }
BasicUsage.argTypes = {
  linkAs: { options: ['link', 'button'], mapping: { link: LinkWithHref } },
}
