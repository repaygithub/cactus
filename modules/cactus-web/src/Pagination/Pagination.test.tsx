import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'

import { StyleProvider } from '../StyleProvider/StyleProvider'
import Pagination from './Pagination'

function ManagedPagination({ size, start, onClick }: any): React.ReactElement {
  const [current, setCurrent] = React.useState(start)
  const callback = !onClick
    ? setCurrent
    : (page: number): void => {
        onClick(page)
        setCurrent(page)
      }
  return <Pagination currentPage={current} pageCount={size} onPageChange={callback} />
}

const expectIcon = (node: HTMLElement, disabled: boolean, label: string): void => {
  expect(node.childNodes).toHaveLength(1)
  expect((node.firstChild as HTMLElement).tagName).toBe('svg')
  expect(node).toHaveAttribute('aria-label', label)
  if (disabled) {
    expect(node).toHaveAttribute('aria-disabled', 'true')
  } else {
    expect(node).not.toHaveAttribute('aria-disabled')
  }
}

const expectPageLink = (node: HTMLElement, page: number, currentPage: number): void => {
  expect(node).toHaveTextContent(`${page}`)
  if (page === currentPage) {
    expect(node).toHaveAttribute('aria-label', `Current page, ${page}`)
    expect(node).toHaveAttribute('aria-disabled', 'true')
    expect(node).toHaveAttribute('aria-current', 'page')
  } else {
    expect(node).toHaveAttribute('aria-label', `Go to page ${page}`)
    expect(node).not.toHaveAttribute('aria-disabled')
    expect(node).not.toHaveAttribute('aria-current')
  }
}

const assertPages = (
  links: HTMLElement[],
  size: number,
  currentPage: number,
  hiddenPages: number[]
): void => {
  const length = size - hiddenPages.length + 4
  expect(links).toHaveLength(length)
  expect(length).toBeLessThan(14)

  expectIcon(links[0], currentPage === 1, 'Go to page 1')
  expectIcon(links[1], currentPage <= 1, `Go to previous page, ${currentPage - 1}`)
  expect(links[1]).toHaveAttribute('rel', 'prev')

  let page = 1
  for (let ix = 2; ix < length - 2; ix++) {
    while (hiddenPages.includes(page)) {
      page += 1
    }
    expectPageLink(links[ix], page++, currentPage)
  }

  expectIcon(links[length - 2], currentPage >= size, `Go to next page, ${currentPage + 1}`)
  expect(links[length - 2]).toHaveAttribute('rel', 'next')
  expectIcon(links[length - 1], currentPage === size, `Go to last page, ${size}`)
}

describe('component: Pagination', (): void => {
  test('11 pages, first page selected', (): void => {
    const { container, getAllByRole } = render(
      <StyleProvider>
        <ManagedPagination size={11} start={1} />
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()

    assertPages(getAllByRole('link'), 11, 1, [8, 9, 10])
  })

  test('11 pages, go to middle page', (): void => {
    const { getAllByRole, getByLabelText } = render(
      <StyleProvider>
        <ManagedPagination size={11} start={1} />
      </StyleProvider>
    )
    userEvent.click(getByLabelText('Go to page 5'))
    assertPages(getAllByRole('link'), 11, 5, [1, 2, 9, 10])
  })

  test('11 pages, go to next page', (): void => {
    const { getAllByRole, getByLabelText } = render(
      <StyleProvider>
        <ManagedPagination size={11} start={5} />
      </StyleProvider>
    )
    userEvent.click(getByLabelText('Go to next page, 6'))
    assertPages(getAllByRole('link'), 11, 6, [1, 2, 3])
  })

  test('11 pages, go to previous page', (): void => {
    const { getAllByRole, getByLabelText } = render(
      <StyleProvider>
        <ManagedPagination size={11} start={5} />
      </StyleProvider>
    )
    userEvent.click(getByLabelText('Go to previous page, 4'))
    assertPages(getAllByRole('link'), 11, 4, [8, 9, 10])
  })

  test('11 pages, go to first page', (): void => {
    const { getAllByRole, getByLabelText } = render(
      <StyleProvider>
        <ManagedPagination size={11} start={5} />
      </StyleProvider>
    )
    userEvent.click(getByLabelText('Go to page 1'))
    assertPages(getAllByRole('link'), 11, 1, [8, 9, 10])
  })

  test('11 pages, go to last page', (): void => {
    const { getAllByRole, getByLabelText } = render(
      <StyleProvider>
        <ManagedPagination size={11} start={5} />
      </StyleProvider>
    )
    userEvent.click(getByLabelText('Go to last page, 11'))
    assertPages(getAllByRole('link'), 11, 11, [1, 2, 3])
  })

  test('9 pages, no hidden pages', (): void => {
    const { getAllByRole, getByLabelText } = render(
      <StyleProvider>
        <ManagedPagination size={9} start={1} />
      </StyleProvider>
    )
    assertPages(getAllByRole('link'), 9, 1, [])
    userEvent.click(getByLabelText('Go to last page, 9'))
    assertPages(getAllByRole('link'), 9, 9, [])
  })

  test('test onclick for disabled buttons', (): void => {
    const func = jest.fn()

    const { getAllByRole } = render(
      <StyleProvider>
        <ManagedPagination size={2} start={1} onClick={func} />
      </StyleProvider>
    )

    const links = getAllByRole('link')
    for (const link of links.slice(0, 3)) {
      userEvent.click(link)
    }
    expect(func).not.toHaveBeenCalled()
    userEvent.click(links[3])
    expect(func).toHaveBeenCalledWith(2)
    for (const link of getAllByRole('link').slice(3)) {
      userEvent.click(link)
    }
    expect(func).toHaveBeenCalledTimes(1)
  })
})
