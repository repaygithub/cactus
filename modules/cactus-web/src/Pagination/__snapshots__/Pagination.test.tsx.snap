// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Pagination 11 pages, first page selected 1`] = `
.c3 {
  vertical-align: middle;
}

.c5 {
  vertical-align: middle;
}

.c4 {
  vertical-align: middle;
  margin-top: 5px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  cursor: default;
  display: block;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c1 {
  min-width: 15px;
  height: 24px;
  text-align: center;
  background: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 2px 8px;
  display: block;
  box-sizing: content-box;
  border-left: 1px solid hsl(200,29%,90%);
}

.c1,
.c1:link {
  color: hsl(200,10%,20%);
  font-size: 15px;
  line-height: 18px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:last-child {
  border-right: 1px solid hsl(200,29%,90%);
}

.c1 svg {
  width: 15px;
  height: 15px;
  vertical-align: middle;
}

.c2 {
  cursor: pointer;
  padding: 3px;
  vertical-align: middle;
  line-height: 18px;
  display: block;
  border-radius: 8px;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  border: none;
  background-color: transparent;
  font: inherit;
}

.c2:visited {
  color: hsl(200,18%,45%);
}

.c2:hover {
  color: hsl(200,96%,35%);
}

.c2:active,
.c2:focus {
  padding: 3px;
  outline: none;
  background-color: hsl(200,96%,35%);
  color: hsl(0,0%,100%);
}

.c2::-moz-focus-inner {
  border: none;
}

.c2[disabled],
.c2[aria-disabled='true'] {
  color: hsl(0,0%,90%);
  background-color: transparent;
  cursor: default;
}

.c2[aria-current='page'] {
  background-color: hsl(200,96%,11%);
  color: hsl(0,0%,100%);
}

@media screen and (min-width:1024px) {
  .c1,
  .c1:link {
    font-size: 15px;
  }
}

<div>
  <nav
    aria-label="Pages"
    class=""
  >
    <ol
      class="c0"
      role="list"
    >
      <li
        class="c1"
      >
        <a
          aria-disabled="true"
          aria-label="Go to page 1"
          class="c2"
          role="link"
        >
          <svg
            fill="currentcolor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.66671 1.33335C2.66671 0.965164 2.36823 0.666687 2.00004 0.666687C1.63185 0.666687 1.33337 0.965164 1.33337 1.33335V14.6667C1.33337 15.0349 1.63185 15.3334 2.00004 15.3334C2.36823 15.3334 2.66671 15.0349 2.66671 14.6667L2.66671 1.33335ZM13.5286 0.861949C13.789 0.6016 14.2111 0.6016 14.4714 0.861949C14.7318 1.1223 14.7318 1.54441 14.4714 1.80476L8.27618 8.00002L14.4714 14.1953C14.7318 14.4556 14.7318 14.8777 14.4714 15.1381C14.2111 15.3984 13.789 15.3984 13.5286 15.1381L6.86197 8.47142C6.60162 8.21107 6.60162 7.78897 6.86197 7.52862L13.5286 0.861949Z"
            />
          </svg>
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-disabled="true"
          aria-label="Go to previous page, 0"
          class="c2"
          rel="prev"
          role="link"
        >
          <svg
            class="c3"
            fill="currentcolor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M17.7071068,2.70710678 C18.0976311,2.31658249 18.0976311,1.68341751 17.7071068,1.29289322 C17.3165825,0.902368927 16.6834175,0.902368927 16.2928932,1.29289322 L6.29289322,11.2928932 C5.90236893,11.6834175 5.90236893,12.3165825 6.29289322,12.7071068 L16.2928932,22.7071068 C16.6834175,23.0976311 17.3165825,23.0976311 17.7071068,22.7071068 C18.0976311,22.3165825 18.0976311,21.6834175 17.7071068,21.2928932 L8.41421356,12 L17.7071068,2.70710678 Z"
            />
          </svg>
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-current="page"
          aria-disabled="true"
          aria-label="Current page, 1"
          class="c2"
          role="link"
        >
          1
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to page 2"
          class="c2"
          role="link"
          tabindex="0"
        >
          2
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to page 3"
          class="c2"
          role="link"
          tabindex="0"
        >
          3
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to page 4"
          class="c2"
          role="link"
          tabindex="0"
        >
          4
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to page 5"
          class="c2"
          role="link"
          tabindex="0"
        >
          5
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to page 6"
          class="c2"
          role="link"
          tabindex="0"
        >
          6
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to page 7"
          class="c2"
          role="link"
          tabindex="0"
        >
          7
        </a>
      </li>
      <li
        class="c1"
      >
        <svg
          class="sc-fznOgF c4"
          fill="currentcolor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M11.75,4.5 C10.7835017,4.5 10,3.71649831 10,2.75 C10,1.78350169 10.7835017,1 11.75,1 C12.7164983,1 13.5,1.78350169 13.5,2.75 C13.5,3.71649831 12.7164983,4.5 11.75,4.5 Z M11.75,13.5 C10.7835017,13.5 10,12.7164983 10,11.75 C10,10.7835017 10.7835017,10 11.75,10 C12.7164983,10 13.5,10.7835017 13.5,11.75 C13.5,12.7164983 12.7164983,13.5 11.75,13.5 Z M11.75,23 C10.7835017,23 10,22.2164983 10,21.25 C10,20.2835017 10.7835017,19.5 11.75,19.5 C12.7164983,19.5 13.5,20.2835017 13.5,21.25 C13.5,22.2164983 12.7164983,23 11.75,23 Z"
          />
        </svg>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to page 11"
          class="c2"
          role="link"
          tabindex="0"
        >
          11
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to next page, 2"
          class="c2"
          rel="next"
          role="link"
          tabindex="0"
        >
          <svg
            class="c5"
            fill="currentcolor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M17.7071068,2.70710678 C18.0976311,2.31658249 18.0976311,1.68341751 17.7071068,1.29289322 C17.3165825,0.902368927 16.6834175,0.902368927 16.2928932,1.29289322 L6.29289322,11.2928932 C5.90236893,11.6834175 5.90236893,12.3165825 6.29289322,12.7071068 L16.2928932,22.7071068 C16.6834175,23.0976311 17.3165825,23.0976311 17.7071068,22.7071068 C18.0976311,22.3165825 18.0976311,21.6834175 17.7071068,21.2928932 L8.41421356,12 L17.7071068,2.70710678 Z"
              transform="matrix(-1 0 0 1 24 0)"
            />
          </svg>
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          aria-label="Go to last page, 11"
          class="c2"
          role="link"
          tabindex="0"
        >
          <svg
            fill="currentcolor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14.6667 1.33335C14.6667 0.965164 14.3682 0.666687 14 0.666687C13.6319 0.666687 13.3334 0.965164 13.3334 1.33335L13.3334 14.6667C13.3334 15.0349 13.6318 15.3334 14 15.3334C14.3682 15.3334 14.6667 15.0349 14.6667 14.6667L14.6667 1.33335ZM1.52864 0.861949C1.26829 1.1223 1.26829 1.54441 1.52864 1.80476L7.7239 8.00002L1.52864 14.1953C1.26829 14.4556 1.26829 14.8777 1.52864 15.1381C1.78899 15.3984 2.2111 15.3984 2.47144 15.1381L9.13811 8.47142C9.39846 8.21107 9.39846 7.78897 9.13811 7.52862L2.47144 0.861949C2.2111 0.6016 1.78899 0.6016 1.52864 0.861949Z"
            />
          </svg>
        </a>
      </li>
    </ol>
  </nav>
</div>
`;
